api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto

```diff
@@ -238,6 +238,62 @@ message HttpConnectionManager {
+  // visible internally if a transformation is enabled. The policy determines
+  // whether Envoy will forward the normalized *:path* header or forward the
+  // original header *:path* header (unless a filter modifies it). Note:
+  // observing (logging, tracing) headers after a transformation is applied will
+  // show the normalized header. Observing after headers are sent to the codec
+  // for encoding to the upstream (e.g. when closing the stream after a response
+  // is received from an upstream) will show the forwarded header.
+  message PathNormalizationOptions {
+    // Defines a policy to apply to a normalization. Any path rewrites that the
+    // router performs (e.g. :ref:`regex_rewrite
+    // <envoy_api_field_config.route.v3.RouteAction.regex_rewrite>` or
+    // :ref:`prefix_rewrite
+    // <envoy_api_field_config.route.v3.RouteAction.prefix_rewrite>`) will apply
+    // to the *:path* header destined for the upstream.
+    enum NormalizationPolicy {
+      // Unspecified. See individual transformations for defaults.
+      UNSPECIFIED = 0;
+
+      // No transformation applied to the header.
+      DISABLE = 1;
+
+      // Normalization applies internally before any processing of requests by
+      // HTTP filters, routing, and matching. Envoy will forward the normalized
+      // *:path* header. Regex rewrites by the router apply to the normalized header.
+      FORWARD = 2;
+
+      // Normalization only applies internally. The original *:path* header will
+      // be restored in the router filter and sent upstream *unless* the path
+      // has been mutated by a filter.
+      INTERNAL = 3;
+    }
+
+    // Should paths be normalized according to RFC 3986 before any processing of
+    // requests by HTTP filters or routing? Defaults to FORWARD (normalization
+    // applies internally and affects upstream *:path* header). When not
+    // specified, this value may be overridden by the runtime variable
+    // :ref:`http_connection_manager.normalize_path<config_http_conn_man_runtime_normalize_path>`. For
+    // paths that fail this check, Envoy will respond with 400 to paths that are
+    // malformed. See `Normalization and Comparison
+    // <https://tools.ietf.org/html/rfc3986#section-6>`_ for details of
+    // normalization. Note that Envoy does not perform `case normalization
+    // <https://tools.ietf.org/html/rfc3986#section-6.2.2.1>`_
+    NormalizationPolicy normalize_path = 1;
+
+    // Determines if adjacent slashes in the path are merged into one before any
+    // processing of requests by HTTP filters or routing. Defaults to
+    // DISABLE. Without setting this option, incoming requests with path
+    // `//dir///file` will not match against route with `prefix` match set to
+    // `/dir`. Note that slash merging is not part of `HTTP spec
+    // <https://tools.ietf.org/html/rfc3986>`_ and is provided for convenience.
+    NormalizationPolicy merge_slashes = 2;
```

Comment:
This might be getting into the implementation weeds, but do you envision any issues if someone selects INTERNAL for one and FORWARD for another? Is it going to lead to complicated state tracking?

Reply:
It's open to do so.

If the `HeaderMap` privately stores the original path, then setting the FORWARD header could be independent of the internal transformed path. (However, we do have to store the original internal header too, to make sure we don't override the path if a filter changed it)

`HeaderMap`:
   * Path: public visible path, HCM sets to INTERNAL path, router sets to FORWARD when forwarding
   * private: `original_path` used to create FORWARD path
   * private: `internal_path` the internal path set by the HCM. used to compare the current path in the router overriding HeaderMap::Path() with FORWARD header.

Reply:
I added a comment about `internal_transformations`:

```
    // These will be applied after :ref:`full_transformation
    // <envoy_api_field_extensions.filters.network.http_connection_manager.v3.HttpConnectionManager.PathNormalizationOptions.full_transformation`
    // is applied. The *:path* header before this transformation will be restored in the router
    // filter and sent upstream unless it was mutated by a filter```

Reply:
@chaoqin-li1123 suggested that we should clarify this too. `full_transformation` (previously called FORWARD) should be though of as incoming header normalization that mutates the header before processing and affects the forwarded.

internal ones only exist temporarily in Envoy during processing. we forward the path with the incoming header normalization.

I am happy to make the name change from `full_transformation` to `incoming_transformation` if that's clearer.